client/js/contextMenuFactory.js

```diff
@@ -0,0 +1,204 @@
+"use strict";
+const $ = require("jquery");
+const socket = require("./socket");
+const utils = require("./utils");
+const JoinChannel = require("./join-channel");
+const ContextMenu = require("./contextMenu");
+const contextMenuActions = [];
+const contextMenuItems = [];
+
+module.exports = {
+	addContextMenuItem,
+	createContextMenu,
+};
+
+addDefaultItems();
+
+/**
+ *
+ * @param opts
+ * @param {function(Object)} [opts.check] - Function to check whether item should show on the context menu, called with the target jquery element, shows if return is truthy
+ * @param {string|function(Object)} opts.className - class name for the menu item, should be prefixed for non-default menu items (if function, called with jquery element, and uses return value)
+ * @param {string|function(Object)} opts.data - data that will be sent to the callback function (if function, called with jquery element, and uses return value)
+ * @param {string|function(Object)} opts.display - text to display on the menu item (if function, called with jquery element, and uses return value)
+ * @param {function(Object)} opts.callback - Function to call when the context menu item is clicked, called with the data requested in opts.data
+ */
+function addContextMenuItem(opts) {
```

Comment:
Do you mind giving the very bare minimal example you expect a package to look like? Like, nothing fancy, doesn't have to be a real use case, but I think writing out how an API gets used is a good way to validate said API.

Reply:
Bear in mind this doesn't include package functionality, but I'm happy to show an example of using the API, sure.

Reply:
You mean there is no way to import a package into a TL instance? If so then that's fine, make all the assumptions or simplifications you need, it's just to make sure we are both looking at the same thing, calibrate expectations, if that makes sense.
(Also a poor-man's documentation for whenever someone stops by and wonders about the API)

Reply:
> You mean there is no way to import a package into a TL instance?

In this PR, yes. That's in #1972.